File: 2asir/web-deployment/PHP/clase/OBJETOS/Python_POO_Mastery_Kit/05_Sistema_Complejo.py
"""
05_Sistema_Complejo.py

OBJETIVO:
El "Factor WOW". Un minijuego RPG de consola.

DIFERENCIAS INTERESANTES:
- rand(min, max): En Python es `random.randint(min, max)`
- echo: `print()`
- Herencia y super(): `super().__init__(...)` para llamar al constructor padre.
"""

import random
from abc import ABC, abstractmethod

# 1. CLASE BASE (ABSTRACTA)
class Personaje(ABC):
    
    def __init__(self, nombre, hp, ataque):
        # En Python, 'protected' es solo convención _variable
        self._nombre = nombre
        self._hp = hp
        self._max_hp = hp
        self._ataque = ataque
        
    def esta_vivo(self):
        return self._hp > 0
    
    def get_nombre(self):
        return self._nombre
    
    def recibir_dano(self, cantidad):
        self._hp -= cantidad
        if self._hp < 0:
            self._hp = 0
        print(f" > {self._nombre} recibe {cantidad} de daño. (HP: {self._hp}/{self._max_hp})")
        
    @abstractmethod
    def atacar(self, objetivo):
        pass

# 2. CLASES HIJAS
class Guerrero(Personaje):
    
    def atacar(self, objetivo):
        min_dmg = self._ataque - 5
        max_dmg = self._ataque + 5
        dano = random.randint(min_dmg, max_dmg)
        
        # Posibilidad de crítico
        if random.randint(1, 10) > 8:
            dano *= 2
            print("¡CRÍTICO BRUTAL!", end=" ")
            
        print(f"⚔️ {self._nombre} golpea con su espada a {objetivo.get_nombre()}!")
        objetivo.recibir_dano(dano)

class Mago(Personaje):
    
    def __init__(self, nombre, hp, ataque):
        # Llamamos al constructor del padre obligatoriamente
        super().__init__(nombre, hp, ataque)
        self.__mana = 50 # Privado real (__)
        
    def atacar(self, objetivo):
        if self.__mana >= 10:
            dano = int(self._ataque * 1.5)
            self.__mana -= 10
            print(f"🔥 {self._nombre} lanza una bola de fuego a {objetivo.get_nombre()}! (Maná: {self.__mana})")
            objetivo.recibir_dano(dano)
        else:
            print(f"💨 {self._nombre} no tiene maná y golpea con su bastón...")
            objetivo.recibir_dano(5)

# 3. MOTOR DEL JUEGO (ESTÁTICO)
class Arena:
    
    @staticmethod
    def combatir(p1, p2):
        print("\n======================================")
        print(f"   COMIENZA EL DUELO: {p1.get_nombre()} vs {p2.get_nombre()}")
        print("======================================\n")
        
        turno = 1
        while p1.esta_vivo() and p2.esta_vivo():
            print(f"\n--- TURNO {turno} ---")
            
            # P1 ataca
            p1.atacar(p2)
            if not p2.esta_vivo(): break
            
            # P2 ataca
            p2.atacar(p1)
            
            turno += 1
            
        Arena.declarar_ganador(p1, p2)
        
    @staticmethod
    def declarar_ganador(p1, p2):
        print("\n======================================")
        if p1.esta_vivo():
            print(f"🏆 ¡VICTORIA PARA {p1.get_nombre()}!")
        else:
            print(f"🏆 ¡VICTORIA PARA {p2.get_nombre()}!")
        print("======================================\n")

## -----------------------------------------
## EJECUCIÓN
## -----------------------------------------

if __name__ == "__main__":
    heroe = Guerrero("Conan", 150, 20)
    villano = Mago("Saruman", 90, 35)
    
    Arena.combatir(heroe, villano)

File: 2asir/web-deployment/PHP/clase/OBJETOS/PHP_POO_Mastery_Kit/05_Sistema_Complejo.php
<?php
/**
 * 05_Sistema_Complejo.php
 * 
 * OBJETIVO:
 * El "Factor WOW". Un minijuego RPG de consola que combina todo lo aprendido.
 * 
 * USA:
 * - Clases, Objetos, Constructores.
 * - Herencia (Personaje -> Guerrero/Mago).
 * - Encapsulamiento (puntos de vida privados).
 * - Lógica de control aleatoria y bucles de juego.
 * - Reporte de estado estático y visual.
 */

// 1. CLASE BASE
abstract class Personaje
{
    protected $nombre;
    protected $hp;       // Puntos de vida
    protected $maxHp;
    protected $ataque;

    public function __construct($nombre, $hp, $ataque)
    {
        $this->nombre = $nombre;
        $this->hp = $hp;
        $this->maxHp = $hp;
        $this->ataque = $ataque;
    }

    public function estaVivo()
    {
        return $this->hp > 0;
    }

    public function getNombre()
    {
        return $this->nombre;
    }

    // Método para recibir daño
    public function recibirDano($cantidad)
    {
        $this->hp -= $cantidad;
        if ($this->hp < 0)
            $this->hp = 0;
        echo " > {$this->nombre} recibe $cantidad de daño. (HP: {$this->hp}/{$this->maxHp})\n";
    }

    // Método abstracto para atacar (cada uno ataca distinto)
    abstract public function atacar(Personaje $objetivo);
}

// 2. CLASES HIJAS (Guerrero y Mago)
class Guerrero extends Personaje
{
    public function atacar(Personaje $objetivo)
    {
        $dano = rand($this->ataque - 5, $this->ataque + 5);
        // Posibilidad de crítico
        if (rand(1, 10) > 8) {
            $dano *= 2;
            echo "¡CRÍTICO BRUTAL! ";
        }
        echo "⚔️ {$this->nombre} golpea con su espada a {$objetivo->getNombre()}!\n";
        $objetivo->recibirDano($dano);
    }
}

class Mago extends Personaje
{
    private $mana = 50;

    public function atacar(Personaje $objetivo)
    {
        if ($this->mana >= 10) {
            $dano = $this->ataque * 1.5; // La magia duele más
            $this->mana -= 10;
            echo "🔥 {$this->nombre} lanza una bola de fuego a {$objetivo->getNombre()}! (Maná: {$this->mana})\n";
            $objetivo->recibirDano($dano);
        } else {
            echo "💨 {$this->nombre} no tiene maná y golpea con su bastón...\n";
            $objetivo->recibirDano(5); // Daño ridículo físico
        }
    }
}

// 3. MOTOR DEL JUEGO (CLASE ESTATICA)
class Arena
{
    public static function combatir(Personaje $p1, Personaje $p2)
    {
        echo "\n======================================\n";
        echo "   COMIENZA EL DUELO: {$p1->getNombre()} vs {$p2->getNombre()}";
        echo "\n======================================\n";

        $turno = 1;
        while ($p1->estaVivo() && $p2->estaVivo()) {
            echo "\n--- TURNO $turno ---\n";

            // Ataca P1
            $p1->atacar($p2);
            if (!$p2->estaVivo())
                break; // Si muere, para

            // Pausa dramática simulada (opcional, quitada por velocidad)

            // Ataca P2
            $p2->atacar($p1);

            $turno++;
        }

        self::declararGanador($p1, $p2);
    }

    private static function declararGanador($p1, $p2)
    {
        echo "\n======================================\n";
        if ($p1->estaVivo()) {
            echo "🏆 ¡VICTORIA PARA {$p1->getNombre()}!";
        } else {
            echo "🏆 ¡VICTORIA PARA {$p2->getNombre()}!";
        }
        echo "\n======================================\n";
    }
}

## -----------------------------------------
## SECCIÓN DE EJECUCIÓN
## -----------------------------------------

$heroe = new Guerrero("Conan", 150, 20); // Mucha vida, ataque medio
$villano = new Mago("Saruman", 90, 35);  // Poca vida, mucho daño

Arena::combatir($heroe, $villano);

/*
 * REFLEXIÓN FINAL:
 * Hemos creado un sistema complejo donde:
 * 1. El motor (Arena) no sabe qué tipos de personajes luchan, solo sabe que son "Personajes".
 * 2. Cada clase gestiona su propia lógica interna (el mago su maná, el guerrero sus críticos).
 * 3. Todo está encapsulado y ordenado. ¡Esto es POO!
 */
?>